source/core_mqtt.c

```diff
@@ -851,6 +852,74 @@ static MQTTStatus_t discardPacket( const MQTTContext_t * pContext,
     return status;
 }
 
+static MQTTStatus_t discardStoredPacket( MQTTContext_t * pContext,
+                                         MQTTPacketInfo_t * pPacketInfo )
+{
+    MQTTStatus_t status = MQTTRecvFailed;
+    int32_t bytesReceived = 0;
+    size_t bytesToReceive = 0U;
+    uint32_t totalBytesReceived = 0U;
+    bool receiveError = false;
+    size_t mqttPacketSize = 0;
+    size_t remainingLength;
+
+    assert( pContext != NULL );
+    assert( pPacketInfo != NULL );
+
+    mqttPacketSize = pPacketInfo->remainingLength + pPacketInfo->headerLength;
+
+    /* Assert that the packet being discarded is bigger than the
+     * receive buffer. */
+    assert( mqttPacketSize > pContext->networkBuffer.size );
+
+    /* Discard these many bytes at a time. */
+    bytesToReceive = pContext->networkBuffer.size;
+
+    /* Number of bytes depicted by 'index' have already been received. */
+    remainingLength = mqttPacketSize - pContext->index;
+
+    while( ( totalBytesReceived < remainingLength ) && ( receiveError == false ) )
+    {
+        if( ( remainingLength - totalBytesReceived ) < bytesToReceive )
+        {
+            bytesToReceive = remainingLength - totalBytesReceived;
+        }
+
+        bytesReceived = recvExact( pContext, bytesToReceive );
```

Comment:
I still don't feel very comfortable with calling recv in a loop here.

Reply:
True, but as we discussed, calling the `recv` function just once means that we'd need to maintain more state as we cannot receive a packet bigger than the buffer in the MQTT library.

So, this is a tradeoff which we can visit later. Does that sound reasonable?

Also, I'll be fixing the failed tests right now.